### PR TITLE
`dipolarkernel`: Implement orientation selection dipolar kernels

### LINKED ISSUE #181: `dipolarkernel`: Excitation bandwidth method is not compatible with Fresnel method yet raises no warnings
If the excitation bandwidth is to be accounted in the dipolar kernel, one would expect: 
````python
K = dl.dipolarkernel(t,r,excbandwidth=wex)
````
to work. However, the excitation bandwidth is only accounted for in the `integral` and `grid` methods for the kernel calculation. Since the default method is `fresnel`, the input is ignored. Only when using 
````python
K = dl.dipolarkernel(t,r,excbandwidth=wex,method='grid') 
````
or
````python
K = dl.dipolarkernel(t,r,excbandwidth=wex,method='integral') 
````
does the value of  `wex` have any effect. 

This must be fixed. In its default behavior, it gives the idea that the value of  `wex` does not have any effect. Either a warning or an error should be raised when specifying the excitation bandwidth and using the Fresnel method. The documentation does also not specify this distiction.

### PR DESCRIPTION
**Pull request by luisfabib**
Introduces a new keyword to `dipolarkernel` that allows passing a distribution of weights for each `theta` value in the powder averaging of the kernel. 

Optimizes the caching of the grid evaluation of the kernel to allow a fast evaluation of the kernel with the same `t` and `r` but different orientation weights. 

Also closes #181.

### CODE REVIEW

File: test/test_dipolarkernel.py [hunk 1]
```diff
@@ -399,3 +399,19 @@ def test_nonuniform_r():
     V0 = K@P
     assert np.round(V0,3) == 1
 #=======================================================================
+
+def test_orisel():
```
Comment (stestoll) [inline test/test_dipolarkernel.py:403]: This is not really a complete test. Add another one where you compare the output of `dipolarkernel` to an (more) explicitly constructed orientationally selective kernel.

File: deerlab/dipolarkernel.py [hunk 2]
```diff
@@ -60,9 +61,15 @@ def dipolarkernel(t, r, *, pathways=None, mod=None, bg=None, method='fresnel', e
...
 
+    orisel : array_like, optional 
+        Probability distribution of possible orientations of the interspin vector to account for orientation selection. Must be 
+        vector defined for uniformly distributed orientations in the range [0,π/2]. If speficied as ``None`` (by default), a uniform distribution is assumed. 
+        Requires the ``'grid'`` method.
+
     excbandwidth : scalar, optional
         Excitation bandwidth of the pulses in MHz to account for limited excitation bandwidth [4]_.
-    
+        Requires the ``'grid'`` or ``'integral'`` methods.
```
Comment (stestoll) [inline deerlab/dipolarkernel.py:71]: It's not implemented for the `integral` method. Please implement, good to have a second decoupled implementation.

File: deerlab/dipolarkernel.py [hunk 3]
```diff
@@ -224,65 +254,71 @@ def calckernelmatrix(t,r,method,wex,nKnots,g):
     nr = np.size(r)
     nt = np.size(t)  
     K = np.zeros((nt,nr))
-    wr = w0(g)/(r**3)  # rad s^-1
+    ωr = ω0(g)/(r**3)  # Mrad s^-1
```
Comment (stestoll) [inline deerlab/dipolarkernel.py:257]: rad µs^-1 is a bit more practically intuitive.

File: deerlab/dipolarkernel.py [hunk 4]
```diff
@@ -224,65 +254,71 @@ def calckernelmatrix(t,r,method,wex,nKnots,g):
...
+
+        if orientationselection:
+            # Interpolate
+            Pθ = interp1d(np.linspace(0,1,int(len(Pθ))),Pθ,kind='cubic')(cosθ)
+            Pθ /= np.sum(Pθ)
+            # Integrate over the orientations distribution Pθ
+            K = np.dot(_Cgrid(ωr,t,ωex,q),Pθ)
+        else: 
+            # Elementary kernel without orientation selection
+            K = np.sum(_Cgrid(ωr,t,ωex,q),2)/nKnots
```
Comment (stestoll) [inline deerlab/dipolarkernel.py:296]: Why is the 2 there? If P0 is isotropic, then it should be 1, right? Or 1/4pi or similar? Check normalization.

File: deerlab/dipolarkernel.py [hunk 5]
```diff
@@ -60,9 +61,15 @@ def dipolarkernel(t, r, *, pathways=None, mod=None, bg=None, method='fresnel', e
         
         The default is ``'fresnel'``.
 
+    orisel : array_like, optional 
+        Probability distribution of possible orientations of the interspin vector to account for orientation selection. Must be 
+        vector defined for uniformly distributed orientations in the range [0,π/2]. If speficied as ``None`` (by default), a uniform distribution is assumed.
```
Comment (stestoll) [inline deerlab/dipolarkernel.py:66]: This is specific to axial distributions, right? P(theta,phi) is not supported? This needs to be stated clearly. Also, describe exactly what theta is.

**Review by stestoll — CHANGES_REQUESTED** [review]

File: deerlab/dipolarkernel.py (hunk 4 above)
Comment (luisfabib) [reply deerlab/dipolarkernel.py:296]: The 2 is the summation axis, in python meaning the third axis, i.e. the dimension along `theta`

File: deerlab/dipolarkernel.py (hunk 5 above)
Comment (luisfabib) [reply deerlab/dipolarkernel.py:66]: Yes, it is specific to axial distributions. Although it would be easy to expand this to include `phi`.

**Comment by luisfabib** [comment]
Looking at the comments I have come with a potentially better implementation. 

- Orientation selection is caused by a non-uniform distribution `P(theta,phi)`. So the functionality should take both `theta` and `phi` into account. 
- The distribution `P(theta,phi)` should be passed as a model function taking `theta` and `phi` as inputs. This would enable the functionality to work with the `method='integral'` mode and in the method, `method='grid'` mode would avoid the need for interpolation to the correct resolution of ``theta``.

More details/commits will follow once I chew this for a bit longer.

File: deerlab/dipolarkernel.py (hunk 2 above)
Comment (luisfabib) [reply deerlab/dipolarkernel.py:73]: Done

File: test/test_dipolarkernel.py (hunk 1 above)
Comment (luisfabib) [reply test/test_dipolarkernel.py]: Added tests against values calculated from Mathematica

File: deerlab/dipolarkernel.py (hunk 3 above)
Comment (luisfabib) [reply deerlab/dipolarkernel.py]: Done

File: deerlab/dipolarkernel.py [hunk 6]
```diff
@@ -213,76 +233,104 @@ def dipolarkernel(t, r, *, pathways=None, mod=None, bg=None, method='fresnel', e
...
-            C = C*np.exp(-(wr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:])**2/wex**2)
-        K = np.sum(C,2)/nKnots
-        return K
+        z = np.linspace(0,1,int(nKnots))
+        θ = np.arccos(z) # rad
+        q = 1 - 3*z**2
+
+        if orientationselection:
+            # Evaluate orientational distribution over grid        
+            Pθnorm,_ = scipy.integrate.quad(lambda z: Pθ(np.arccos(z)),0,1,limit=1000)
```
Comment (stestoll) [inline deerlab/dipolarkernel.py:297]: Based on symmetry, this interpolant must have zero derivative at 0 and pi/2

**Review by stestoll — CHANGES_REQUESTED** [review]

File: deerlab/dipolarkernel.py (hunk 5 above)
Comment (stestoll) [reply deerlab/dipolarkernel.py]: I think only theta matters for the DEER signal. The user should take care of integration over phi.

**Comment by luisfabib** [comment]
The last set of commits introduces an important speedup to the `method='integral'` mode (about 100 times faster). This makes the method more competitive with respect to the `method='grid'`.  

Example: Elementary dipolar kernel `t`: 1000 points `r`: 500 points 
- Before: 239s
- Now: 2s

File: deerlab/dipolarkernel.py (hunk 6 above)
Comment (luisfabib) [reply deerlab/dipolarkernel.py]: It was not strictly satisfied. It depended on whether the user enforced that on the input function. Now `dipolarkernel` constructs a constrained spline from the input distribution to ensure zero derivatives at 0 and pi/2.

**Review by stestoll — APPROVED** [review]

**Pull request merged**